vsintegration/src/FSharp.Editor/Common/LanguageService.fs

```diff
@@ -245,32 +242,59 @@ and
     internal FSharpLanguageService(package : FSharpPackage) =
     inherit AbstractLanguageService<FSharpPackage, FSharpLanguageService>(package)
 
-    let checkerProvider = package.ComponentModel.DefaultExportProvider.GetExport<FSharpCheckerProvider>().Value
     let projectInfoManager = package.ComponentModel.DefaultExportProvider.GetExport<ProjectInfoManager>().Value
 
     let projectDisplayNameOf projectFileName = 
         if String.IsNullOrWhiteSpace projectFileName then projectFileName
         else Path.GetFileNameWithoutExtension projectFileName
 
-    let openedProjects = Queue<IProvideProjectSite>()
+    let singleFileProjects = ConcurrentDictionary<_, AbstractProject>()
+
+    let tryRemoveSingleFileProject projectId =
+        match singleFileProjects.TryRemove(projectId) with
+        | true, project ->
+            projectInfoManager.RemoveSingleFileProject(projectId)
+            project.Disconnect()
+        | _ -> ()
+
+    override this.Initialize() =
+        base.Initialize()
 
-    do
-      Events.SolutionEvents.OnAfterOpenProject.Add (fun (args: Events.OpenProjectEventArgs) -> 
-        match args.Hierarchy with
-        | :? IProvideProjectSite as siteProvider ->
-            openedProjects.Enqueue(siteProvider)            
-        | _ -> ())
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Completion.CompletionOptions.BlockForCompletionItems, FSharpCommonConstants.FSharpLanguageName, false)
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Shared.Options.ServiceFeatureOnOffOptions.ClosedFileDiagnostic, FSharpCommonConstants.FSharpLanguageName, Nullable false)
+
+        this.Workspace.DocumentClosed.Add <| fun args ->
+            tryRemoveSingleFileProject args.Document.Project.Id 
+            
+        Events.SolutionEvents.OnAfterCloseSolution.Add <| fun _ ->
+            singleFileProjects.Keys |> Seq.iter tryRemoveSingleFileProject
+
+        let ctx = System.Threading.SynchronizationContext.Current
         
-    member this.BatchSetupProjects() =    
-        if openedProjects.Count > 0 then    
-            let workspace = package.ComponentModel.GetService<VisualStudioWorkspaceImpl>()
-            for siteProvider in openedProjects do
-                this.SetupProjectFile(siteProvider, workspace)
-            openedProjects.Clear()
+        let rec setupProjectsAfterSolutionOpen() =
+            async {
+                use openedProjects = MailboxProcessor.Start <| fun inbox ->
+                    async { 
+                        // waits for AfterOpenSolution and then starts projects setup
+                        do! Async.AwaitEvent Events.SolutionEvents.OnAfterOpenSolution |> Async.Ignore
+                        while true do
+                            let! siteProvider = inbox.Receive()
+                            do! Async.SwitchToContext ctx
```

Comment:
Or can `Async.SwitchToContext` be moved outside of the loop?
I always thought it "sticks" only till the next `let!` or `do!` but it seems not to be the case.